README.md
`hode` is a reimplementation of the engine used by the game 'Heart of Darkness'
developed by Amazing Studio.

Datafiles
=========
The original datafiles from the Windows releases (Demo or CD) are required.

- hod.paf (hod_demo.paf, hod_demo2.paf)
- setup.dat
- *_hod.lvl
- *_hod.sss
- *_hod.mst

See also the `RELEASES.yaml` file for a list of game versions this program
has been tested with.

Compiling
=========
Windows
-------
```
git clone --recurse-submodules https://github.com/CommonLoon102/hode-vs
cd hode-vs
3p\vcpkg\bootstrap-vcpkg.bat
cmake -B build -DCMAKE_BUILD_TYPE=Release -A x64 -DVCPKG_TARGET_TRIPLET=x64-windows -DCMAKE_TOOLCHAIN_FILE=%CD%\3p\vcpkg\scripts\buildsystems\vcpkg.cmake
cmake --build build --config Release --parallel 3
```

Linux
-----
```
sudo apt update
sudo apt install libsdl2-dev git make cmake gcc g++ -y
git clone --recurse-submodules https://github.com/CommonLoon102/hode-vs
cd hode-vs
cmake -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build --config Release --parallel 3
```

macOS
-----
```
brew upgrade
brew install sdl2
git clone --recurse-submodules https://github.com/CommonLoon102/hode-vs
cd hode-vs
cmake -B build -DCMAKE_BUILD_TYPE=Release
cmake --build build --config Release --parallel 3
```

Debugging
=========
Windows
-------
```
git clone --recurse-submodules https://github.com/CommonLoon102/hode-vs
cd hode-vs
3p\vcpkg\bootstrap-vcpkg.bat
```
1. **As admin** in the `hode-vs\3p\vcpkg` folder: `vcpkg.exe integrate install`
2. Open the `hode-vs` folder in Visual Studio 2019 (Get started -> Open a local folder)
3. Wait until CMake installs the packages via vcpkg
4. Next to the play button, select startup item: `hode.exe`
5. Click play
6. Copy the needed files described in the `Datafiles` section into the debug folder

Running
=======
By default the engine will try to load the files from the current directory
and start the game from the first level.

These defaults can be changed using command line switches :

    Usage: hode [OPTIONS]...
    --datapath=PATH   Path to data files (default '.')
    --savepath=PATH   Path to save files (default '.')
    --level=NUM       Start at level NUM
    --checkpoint=NUM  Start at checkpoint NUM

Display and engine settings can be configured in the `hode.ini` file.

Game progress is saved in `setup.cfg`, similar to the original engine.

Credits
=======
All the team at Amazing Studio for possibly the best cinematic platformer ever
developed.

Contact
=======
Gregory Montoir, cyx@users.sourceforge.net

URLs
====
- https://www.mobygames.com/game/heart-of-darkness
- http://heartofdarkness.ca/


CHANGES.txt
* release 0.2.9f
	- added 'linear' graphics scaler
	- enabled PSX options menu
	- fixed '1x' scaling

* release 0.2.9e
	- fixed graphics glitches with PSX backgrounds
	- removed 'inih' and 'libxbr-standalone' dependencies

* release 0.2.9d
	- added gain to sounds
	- added no-hit cheat for 'rock' and 'lava'

* release 0.2.9c
	- added original sounds limiter (16 channels)
	- fixed checkpoint 7 for 'lar1' screen 19
	- fixed some bounding box checks
	- fixed sound glitches when resuming the game after a cutscene

* release 0.2.9b
	- added PSX hint screens
	- added 'volume' submenu
	- fixed sounds volume and panning
	- fixed walk on chain bridge in 'lar1' screen 3
	- fixed gate not closing in 'lar2' screen 8

* release 0.2.9
	- added PSP and Wii platforms specific code
	- added support for PC demo v1.0 data files
	- fixed screen state for 'fort' screens 16 and 17
	- fixed checkpoint for 'pwr1' screen 21
	- fixed PAF animation glitches with last frame

* release 0.2.8
	- added PSX background overlays (MDEC)
	- fixed crash playing PAF animation #3 with Italian PC data files

* release 0.2.7
	- added 'projection' submenu
	- added PSX backgrounds (MDEC)
	- fixed menu on big endian platforms

* release 0.2.6
	- added initial code for menu
	- added PSX sounds (SPU ADPCM)
	- fixed skull animation in 'rock' screen 18

* release 0.2.5
	- fixed spiders and worms movement
	- several fixes for big endian platforms

* release 0.2.4
	- fixed Amigo walk path in 'pwr2' screen 7
	- fixed glowing stone puzzle in 'lar2'
	- fixed missing trigger for switch #0 in 'lar1'

* release 0.2.3
	- fixed double spectre fireball animations in 'lar1' and 'lar2'
	- fixed level transition from 'pwr1' to 'isld'
	- fixed screen state for 'lava' screen 15

* release 0.2.2
	- fixed lever in 'lava' level screen 2
	- fixed masking with plants in 'pwr1' level
	- fixed missing explosion animation with fireballs
	- fixed screen mask in 'lar2' level screen 7

* release 0.2.1
	- added initial code for PSX data files detection
	- fixed sprite infinite looping in 'pwr1' level screen 23
	- several fixes to gates and switches in 'lar1' and 'lar2' levels

* release 0.2.0
	- added missing postScreenUpdate callbacks for 'lava' level
	- fixed crash at beginning of 'lar1' level
	- fixed gamma ramp with widescreen borders
	- several fixes to gates handling in 'lar1' and 'lar2' levels

* release 0.1.9
	- fixed fireballs not hitting Andy
	- fixed fishes movement in 'pwr1' and 'pwr2' levels
	- fixed several bounding box clipping checks
	- fixed sounds panning

* release 0.1.8
	- added specific callbacks for 'lar1' and 'lar2' levels
	- fixed infinite loop in .mst op27
	- several fixes to sound code

* release 0.1.7
	- added 16:9 widescreen (blur)
	- added missing preScreenUpdate callbacks for 'lar1' level
	- fixed .sss version 12 loading (demo v1.4)

* release 0.1.6
	- fixed Andy animation glitches when climbing
	- fixed Andy apnea animation in 'pwr1' level
	- fixed duplicated cutscene playback when switching levels
	- fixed postScreenUpdate callback for 'isld' level screen 15
	- fixed postScreenUpdate callback for 'pwr1' level screen 6

* release 0.1.5
	- added 'firefly' object in 'fort' level
	- added collision with monsters
	- added missing explosion sprites with plasma cannon
	- added postScreenUpdate callbacks for 'isld' level
	- fixed monsters hug

* release 0.1.4
	- added detection for sector files (demo)
	- fixed hint screens displaying too often
	- fixed monsters initial position
	- fixed skull animation in 'rock' level

* release 0.1.3
	- added initial code for 'Special Powers'
	- implemented .mst opcodes for 'rock' level
	- fixed palette fading when dying

* release 0.1.2
	- added .mst screen area triggers
	- added parsing for .mst files
	- added postScreenUpdate callbacks for 'lar1' level

* release 0.1.1
	- added audio playback for PAF animations
	- added postScreenUpdate callbacks for 'lava' level
	- added preScreenUpdate callbacks for 'isld' level
	- added transform screen tables for levels 'pwr1' and 'pwr2'
	- fixed timing for PAF animations (10 fps)

* release 0.1.0
	- first release


paf.h
/*
 * Heart of Darkness engine rewrite
 * Copyright (C) 2009-2011 Gregory Montoir (cyx@users.sourceforge.net)
 */

#ifndef PAF_PLAYER_H__
#define PAF_PLAYER_H__

#include "intern.h"
#include "defs.h"
#include "fileio.h"

struct PafHeader {
	uint32_t preloadFrameBlocksCount;
	uint32_t *frameBlocksCountTable;
	uint32_t *framesOffsetTable;
	uint32_t *frameBlocksOffsetTable;
	int32_t framesCount;
	int32_t frameBlocksCount;
	uint32_t startOffset;
	uint32_t readBufferSize;
	int32_t maxVideoFrameBlocksCount;
	int32_t maxAudioFrameBlocksCount;
	int32_t frameDuration;
};

// names taken from the PSX filenames
enum {
	kPafAnimation_intro = 0,
	kPafAnimation_cine14l = 1,
	kPafAnimation_rapt = 2,
	kPafAnimation_glisse = 3,
	kPafAnimation_meeting = 4,
	kPafAnimation_island = 5,
	kPafAnimation_islefall = 6,
	kPafAnimation_vicious = 7,
	kPafAnimation_together = 8,
	kPafAnimation_power = 9,
	kPafAnimation_back = 10,
	kPafAnimation_dogfree1 = 11,
	kPafAnimation_dogfree2 = 12,
	kPafAnimation_meteor = 13,
	kPafAnimation_cookie = 14,
	kPafAnimation_plot = 15,
	kPafAnimation_puzzle = 16,
	kPafAnimation_lstpiece = 17,
	kPafAnimation_dogfall = 18,
	kPafAnimation_lastfall = 19,
	kPafAnimation_end = 20,
	kPafAnimation_cinema = 21,
	kPafAnimation_CanyonAndyFallingCannon = 22, // kPafAnimation_ghct
	kPafAnimation_CanyonAndyFalling = 23, // kPafAnimation_chute
	kPafAnimation_IslandAndyFalling = 24 // kPafAnimation_chute_i
};

struct FileSystem;

struct PafAudioQueue {
	int16_t *buffer; // stereo samples
	int offset, size;
	PafAudioQueue *next;
};

struct PafCallback {
	void (*frameProc)(void *userdata, int num, const uint8_t *frame);
	void (*endProc)(void *userdata);
	void *userdata;
};

struct PafPlayer {

	enum {
		kMaxVideosCount = 50,
		kBufferBlockSize = 2048,
		kVideoWidth = 256,
		kVideoHeight = 192,
		kPageBufferSize = 256 * 256,
		kAudioSamples = 2205,
		kAudioStrideSize = 4922 // 256 * sizeof(int16_t) + 2205 * 2
	};

	bool _skipCutscenes;
	FileSystem *_fs;
	File _file;
	int _videoNum;
	uint32_t _videoOffset;
	PafHeader _pafHdr;
	int _currentPageBuffer;
	uint8_t *_pageBuffers[4];
	uint8_t _paletteBuffer[256 * 3];
	bool _paletteChanged;
	uint8_t _bufferBlock[kBufferBlockSize];
	uint8_t *_demuxVideoFrameBlocks;
	uint8_t *_demuxAudioFrameBlocks;
	uint32_t _audioBufferOffsetRd;
	uint32_t _audioBufferOffsetWr;
	PafAudioQueue *_audioQueue, *_audioQueueTail;
	uint32_t _flushAudioSize;
	uint32_t _playedMask;
	PafCallback _pafCb;
	int _volume;
	int _frameMs;

	PafPlayer(FileSystem *fs);
	~PafPlayer();

	void setVolume(int volume);

	void preload(int num);
	void play(int num);
	void unload(int num = -1);

	bool readPafHeader();
	uint32_t *readPafHeaderTable(int count);

	void decodeVideoFrame(const uint8_t *src);
	uint8_t *getVideoPageOffset(uint16_t val);
	void decodeVideoFrameOp0(const uint8_t *base, const uint8_t *src, uint8_t code);
	void decodeVideoFrameOp1(const uint8_t *src);
	void decodeVideoFrameOp2(const uint8_t *src);
	void decodeVideoFrameOp4(const uint8_t *src);

	void decodeAudioFrame(const uint8_t *src, uint32_t offset, uint32_t size);

	void mix(int16_t *buf, int samples);
	void mainLoop();

	void setCallback(const PafCallback *pafCb);
};

#endif // PAF_PLAYER_H__
